### PR TITLE
Remove the voter election process

### PR DESCRIPTION
**Pull request by ulbqb**
# Abstract

This PR remove the voter election process from Ostracon. 

# Background

Ostracon need some compatibility with Tendermint for IBC and cosmos-sdk.  It it necessary to change possible parts for compatibility. Currently the voter election process is executed in Ostracon. Since voter selection can be moved to the lbm-sdk layer, it want to be moved to the lbm-sdk layer without putting it in Ostracon.

# Goal

The goal is to fix to be compatible with Tendermint about voter election process. This means that the voter election process is removed from ostracon and it is done by only abci app. This goal need the following changes.

- Remove the voter election process
- Remove the proto field related to voter election 
  - Remove`voter_params`, `voter`, `voters`, `last_voters`, `voters_hash`, `voter_set`, `voting_weight` that does not exist in Tendermint.
  - `voter_params` is the parameters used by voter election so this is not needed.
  - `voters` is the nodes elected from validators so this is not needed.
  - `voting_weight` is the value recalculated as voting power in voters so this is not needed. 
- Remove RPCs related to voter
  - `Voters`, `ValidatorsWithVoters`
- FIx openapi.yaml
- Fix documents

# Responsibility
The changes is based on Tendermint v0.34.19. 
This PR contains fixes other than documentation fixes. Fixing the documents is in #544. 

# Results
 The voter election process was removed from ostracon. However, voter selection can still be implemented with the abci app. This PR makes it impossible to change the validator for each round. Updating validators is perform in EndBlock. See [tendermint docs](https://docs.tendermint.com/v0.33/app-dev/app-development.html#endblock) for details.

### CODE REVIEW

File: test/e2e/tests/validator_test.go [hunk 1]
```diff
@@ -56,17 +56,6 @@ func TestValidator_Sets(t *testing.T) {
 					"incorrect PubKey of validator set at height %v", h)
 				require.Equal(t, expected.VotingPower, actual.VotingPower,
 					"incorrect VotingPower of validator set at height %v", h)
-				if node.Name == "full02" {
```
Comment (torao) [inline test/e2e/tests/validator_test.go:59]: What's the reason why this step could be better to be deleted in this PR?

**Comment by torao** [comment]
I can then accept this PR if the details of the backgrounds and objectives of this modification are written.

File: proto/ostracon/abci/types.proto [hunk 2]
```diff
@@ -400,7 +397,7 @@ message Evidence {
   // Total voting power of the validator set in case the ABCI application does
   // not store historical validators.
   // https://github.com/tendermint/tendermint/issues/4581
-  int64 total_voting_power = 5; // total voting weight
+  int64 total_voting_power = 5;  // total voting weight
```
Comment (tnasu) [inline proto/ostracon/abci/types.proto:400]:
```suggestion
  int64 total_voting_power = 5;  // total voting power
```

**Review by tnasu — CHANGES_REQUESTED** [review]

File: proto/ostracon/types/evidence.proto [hunk 3]
```diff
@@ -19,20 +19,17 @@ message Evidence {
...
-  int64 voting_weight = 1000;
 }
 
 // LightClientAttackEvidence contains evidence of a set of validators attempting to mislead a light client.
 message LightClientAttackEvidence {
   ostracon.types.LightBlock conflicting_block            = 1;
   int64                     common_height                = 2;
   repeated ostracon.types.Validator byzantine_validators = 3;
-  int64                             total_voting_power   = 4; // total voting weight
+  int64                             total_voting_power   = 4;  // total voting weight
```
Comment (tnasu) [inline proto/ostracon/types/evidence.proto:32]:
```suggestion
  int64                             total_voting_power   = 4;  // total voting power
```

File: consensus/reactor_test.go [hunk 4]
```diff
@@ -318,12 +318,12 @@ func TestReactorRecordsVotesAndBlockParts(t *testing.T) {
 //-------------------------------------------------------------
 // ensure we can make blocks despite cycling a validator set
 
-func TestReactorVotingWeightChange(t *testing.T) {
+func TestReactorVotingPowerChange(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ulbqb) [reply consensus/reactor_test.go:321]: `voting_weight` is the value recalculated as voting power in voters so this is not needed.

**Comment by ulbqb** [comment]
@torao 
I updated PR description. Please comment if anything is missing.

**Review by tnasu — APPROVED** [review]

**Comment by torao** [comment]
@ulbqb 
Does LINE Blockchain no longer support Voter elections? If so, how about the reward payout would? Or if not, how to implement voter election? This should be important information for product design.

**Comment by zemyblue** [comment]
> @ulbqb
> Does LINE Blockchain no longer support Voter elections? If so, how about the reward payout would? Or if not, how to implement voter election? This should be important information for product design.

@torao , we need to move this voter selection function to lbm-sdk layer later.

**Review by zemyblue — COMMENTED** [review]
> Currently the voter election process is executed in Ostracon. However, if Ostracon's spec follows Tendermint's spec for this process, Ostracon should not have vote election process. Because Tendermint is not responsible for voter election, it is done by abci app.

@ulbqb , I think it is not a problem of description. 
Ostracon need some compatibility with Tendermint for IBC and cosmos-sdk. So I think this PR is necessary because it is necessary to change possible parts for compatibility, and since voter selection can be moved to the lbm-sdk layer, it is moved to the lbm-sdk layer without putting it in Ostracon.

**Comment by ulbqb** [comment]
> @ulbqb
> Does LINE Blockchain no longer support Voter elections? If so, how about the reward payout would? Or if not, how to implement voter election? This should be important information for product design.

Voter selection and reward logic can be freely implemented with abci app(lbm-sdk). Updating validators is perform in EndBlock. See [tendermint docs](https://docs.tendermint.com/v0.33/app-dev/app-development.html#endblock) for details.

File: abci/client/mocks/client.go [hunk 5]
```diff
@@ -837,3 +837,18 @@ func (_m *Client) String() string {
 
 	return r0
 }
+
+type mockConstructorTestingTNewClient interface {
```
Comment (kokeshiM0chi) [inline abci/client/mocks/client.go:841]: I would like to know why this was necessary.

**Review by kokeshiM0chi — APPROVED** [review]
LGTM

**Comment by torao** [comment]
@zemyblue @ulbqb All right. We have an implicit agreement in our private discussion about how to conduct voter elections, but this PR only contains bits and pieces of information, which is why I asked the question.

To state our conclusion for those interested in LINE Blockchain, we plan to move the behaviour known as *voter election* from `ostracon` to the `lbm-sdk`. This is because in Tendermint, the selection of Validators is the responsibility of the Cosmos-SDK and we have chosen to follow that approach. In this PR, we'll remove the voter election from Ostracon and implement it in a future update to the `lbm-sdk.

File: test/e2e/tests/validator_test.go (hunk 1 above)
Comment (torao) [reply test/e2e/tests/validator_test.go:59]: I see.

File: consensus/reactor_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (torao) [reply consensus/reactor_test.go:321]: Ok

**Review by torao — APPROVED** [review]

File: abci/client/mocks/client.go (hunk 5 above)
Comment (ulbqb) [reply abci/client/mocks/client.go:841]: This was generated automatically by mockery.

**Comment by egonspace** [comment]
Hi guys~
Wow, this is amazing! 
If so, do you give up functions such as selecting proposer according to probability? I don't think the sdk layer can be involved in selecting the proposer.

**Comment by ulbqb** [comment]
@egonspace 
Hello.

> If so, do you give up functions such as selecting proposer according to probability? 

No. This PR removed only the voter election from ostracon. The proposer election is still run by ostracon.

**Pull request merged**